### PR TITLE
Added `ArraySegmentTree` and `RangeQueryDynamic`

### PR DESCRIPTION
**Pull request by Rajveer100**
Added Segment Trees with Lazy Propagation :

Functions : RangeQuery, RangeUpdate, PointUpdate...(O(log(n)) time..

Will be further adding more additions for better user experience...

### CODE REVIEW

**Comment by czgdp1807** [comment]
I would name it as `DynamicSegmentTree` and keep it in a new file, `miscellaneous_data_structures/dynamic_segment_tree.py`. In addition, please care to add some tests and documentation strings. Take some time to understand the coding pattern inside the project before adding your code.

**Comment by Rajveer100** [comment]
Thank you for your feedback..I will make these changes..as soon as possible to make it more readable...
👍

**Comment by Rajveer100** [comment]
Hello @czgdp1807, I have added all necessary DocStrings with necessary examples and test cases for using the data structure...And have also placed it in the right folder [miscellaneous data structures] as you mentioned..

I request you to review my updated code which I have committed..

File: pydatastructs/trees/segmentTree.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+class SegmentTree :
```
Comment (czgdp1807) [inline pydatastructs/trees/segmentTree.py:1]: Remove this file.

File: pydatastructs/miscellaneous_data_structures/segment_tree.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+            _node_ //= 2
+
+    """
+    -- All functions below are the User Functions which invoke the necessary
+       Internal Functions as mentioned earlier. --
+    """
+
+    def rangeUpdate(self, _left_, _right_, _newVal_) :
+
+        self.__rangeUpdate__(1, 0, self._treeSize_ // 2 - 1, _left_, _right_, _newVal_)
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/segment_tree.py:212]: Doesn't make sense to have a method having `__` in the prefix and suffix. These are only used for magic methods in Python.

File: pydatastructs/miscellaneous_data_structures/segment_tree.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    def pointUpdate(self, _pos_, _newVal_) :
+
+        self.__pointUpdate__(_pos_, _newVal_)
+
+    def rangeQuery(self, _left_, _right_) :
+
+        return self.__rangeQuery__(1, 0, self._treeSize_ // 2 - 1, _left_, _right_)
+
+    """
+    Example for using the SegmentTree class (with testcase) :
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/segment_tree.py:223]: By tests I meant in a file like, `test_segment_tree.py`. Would suggest to take a look at `tests` folders in the project. You will get to know about writing tests for an API in Python.

File: pydatastructs/trees/segmentTree.py (hunk 1 above)
Comment (Rajveer100) [reply pydatastructs/trees/segmentTree.py]: Yes, removed it..

File: pydatastructs/miscellaneous_data_structures/segment_tree.py (hunk 2 above)
Comment (Rajveer100) [reply pydatastructs/miscellaneous_data_structures/segment_tree.py:212]: Ok..what name would you prefer?

File: pydatastructs/miscellaneous_data_structures/segment_tree.py (hunk 3 above)
Comment (Rajveer100) [reply pydatastructs/miscellaneous_data_structures/segment_tree.py:223]: Oh...Alright...will add that..

**Comment by Rajveer100** [comment]
> Overall, you have worked hard on it. Though, have you included n-dimensional segment trees? I think that would make this feature applicable to a lot more use cases.

Thank you so much for your respect...I haven't added n-dimensional tree yet..and I actually request you to give me little time for that..but would be obliged..if you could merge my project after I add the test cases and also make the changes as you mentioned above...Thank You..Will commit the changes as you said above..

File: AUTHORS [hunk 4]
```diff
@@ -1,11 +0,0 @@
...
-Kartikei Mittal <kartikeimittal@gmail.com>
-Umesh <23umesh.here@gmail.com>
-Rohan Singh <singh.77@iitj.ac.in>
-Tarun Singh Tomar <tomartarun2001@gmail.com>
-Saptashrungi Birajdar <saptashrungib@gmail.com>
-Rajiv Ranjan Singh <rajivperfect007@gmail.com>
-Prashant Rawat <prashant.rawat216@gmail.com>
-Harsheet <harsheetkakar@gmail.com>
-Pratik Goyal <pratikgoyal2712@gmail.com>
-Jay Thorat <j.thorat10@gmail.com>
```
Comment (czgdp1807) [inline AUTHORS:11]: Please don't remove AUTHORS file. 😅

Comment (Rajveer100) [reply AUTHORS:11]: Fixed..

**Comment by czgdp1807** [comment]
Do you have time for a Google meet?

**Comment by Rajveer100** [comment]
Would be grateful!...What time?

**Comment by czgdp1807** [comment]
Can you please send you email ID in discord DM?

**Comment by Rajveer100** [comment]
I have emailed you the details..

**Comment by czgdp1807** [comment]
None received. Please DM me on discord. Thanks.

**Comment by Rajveer100** [comment]
I don't have your discord id...

**Comment by czgdp1807** [comment]
It's czgdp1807#5025

**Comment by Rajveer100** [comment]
My discord id : Rajveer#2867

**Comment by Rajveer100** [comment]
Have added the tests..

**Comment by czgdp1807** [comment]
The code looks really good. Can you make changes to the tests as well in a similar manner (as we discussed on our call)?

@Smit-create Can you please check the implementation for correctness? Any API suggestions from your experience?

**Comment by Rajveer100** [comment]
Thanks..I will do the commit for the tests soon..

**Comment by Rajveer100** [comment]
Added the updated tests..as well...please review it.

**Comment by Rajveer100** [comment]
Added the updated tests..as well...please review it.

**Comment by Rajveer100** [comment]
Corrected header file directory...Please check the commit..

File: pydatastructs/miscellaneous_data_structures/tests/test_segment_trees.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
+import random
+from pydatastructs.miscellaneous_data_structures.segment_tree import SegmentTree
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/tests/test_segment_trees.py:2]: That's here the problem is. Look for `__all__` variables. Try to observe the pattern on how imports happen via, `from pydatastructs import some_stuff`.

Comment (Rajveer100) [reply pydatastructs/miscellaneous_data_structures/tests/test_segment_trees.py:2]: Alright...

**Comment by Rajveer100** [comment]
I think now it's fine...

**Comment by Rajveer100** [comment]
Hopefully this works now...

**Comment by Rajveer100** [comment]
Could you please check now...

**Comment by Rajveer100** [comment]
Could you review it again?

**Comment by Rajveer100** [comment]
Could you review it, I have corrected the white spaces and new line error...

**Comment by czgdp1807** [comment]
I will review tomorrow.

**Comment by Rajveer100** [comment]
You were going to review today right?

**Comment by Rajveer100** [comment]
Hey, waiting for your review..Thanks.

**Comment by Rajveer100** [comment]
Hey there, waiting for your review..

**Comment by czgdp1807** [comment]
Hi. Please be patient. Meanwhile, feel free to start working on another feature. :-)

**Comment by Rajveer100** [comment]
Yeah I had got you the first time..Lol..Thanks anyway..for replying me..Grateful for that..Yes..I have already started working on the n dimensional part..

File: pydatastructs/miscellaneous_data_structures/segment_tree.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+                self.tree[self.tree_size // 2 + i] = init_list[i]
+
+            for i in range(list_size, self.tree_size // 2):
+
+                self.tree[self.tree_size // 2 + i] = 0
+
+            for i in range(self.tree_size // 2 - 1, 0, -1):
+
+                self.tree[i] = self.tree[2 * i] + self.tree[2 * i + 1]
```
Comment (Smit-create) [inline pydatastructs/miscellaneous_data_structures/segment_tree.py:95]: This is just for additions right? How about taking the custom merge functions with 2 arguments while initialising the tree?.
Like: 
`f = lambda a, b: a + b`

File: pydatastructs/miscellaneous_data_structures/tests/test_segment_trees.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+def random_test():
+    """
+    Random test for the Segment Tree using random list
+    of values, random range queries.
+    """
+    random_list = [random.randint(-10 ** 6, 10 ** 6) for _ in range(10 ** 5)]
+    t = SegmentTree(random_list)
+
+    for _ in range(random.randint(10 ** 3, 10 ** 6)):
```
Comment (Smit-create) [inline pydatastructs/miscellaneous_data_structures/tests/test_segment_trees.py:23]: We don't need to test this for such large values. It consumes much time on CI unnecessarily (python being slower).

File: pydatastructs/miscellaneous_data_structures/segment_tree.py (hunk 6 above)
Comment (Rajveer100) [reply pydatastructs/miscellaneous_data_structures/segment_tree.py:95]: Yes, I had known about it..will add it..

File: pydatastructs/miscellaneous_data_structures/segment_tree.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+    https://cp-algorithms.com/data_structures/segment_tree.html
+    """
+    def __init__(self, inp_list):
+
+        def get_len(size):
+            """
+            Returns the new list size (the closest power of 2).
+
+            Parameters
```
[the remark this answers is not in the document]
Comment (Rajveer100) [reply pydatastructs/miscellaneous_data_structures/segment_tree.py:56]: Reason?

File: pydatastructs/miscellaneous_data_structures/tests/test_segment_trees.py (hunk 7 above)
Comment (Rajveer100) [reply pydatastructs/miscellaneous_data_structures/tests/test_segment_trees.py:23]: Yes, that's the reason I have commented it out..haven't actually called the function..

**Comment by czgdp1807** [comment]
Now, we need to add `RangeQueryDynamic` and `RangeQueryDynamicOneDimensionalArraySegmentTree` in `pydatastructs/miscellaneous_data_structures/algorithms.py`.

Lazy propagation can be added afterwards by adding a new class, something like, `OneDimensionalArraySegmentTreeWithLazyPropagation`. The reason is simple, we would need a new kind of node, i.e., `LazyTreeNode` which would have an attribute to store whether this node is required to be updated. I would recommend doing this in a follow up PR.

**Comment by Rajveer100** [comment]
Could you explain be in detail, of all the changes you did above..as it was quite a lot..

**Comment by Rajveer100** [comment]
Have gone through the commits..but I needed your words..

**Comment by czgdp1807** [comment]
I decoupled the interface and the implementation of segment trees. Basically, segment trees with range updates need to track some extra information and require extra memory, so it would be better to add them separately in a follow up PR. Storing segment trees in arrays would require 4*N sized array which is too much provided the number of nodes are always constant in a segment tree.

File: pydatastructs/miscellaneous_data_structures/segment_tree.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                    parent_node = recursion_stack.pop()
+                    recursion_stack.push((parent_node.key[0], node))
+            else:
+                mid = (start + end) // 2
+                if start <= index and index <= mid:
+                    recursion_stack.push((node.left, None))
+                else:
+                    recursion_stack.push((node.right, None))
+
+    def _query(self, node, start, end, l, r):
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/segment_tree.py:132]: `if not self.is_ready` to be added here.

**Comment by Rajveer100** [comment]
My segment tree takes 2*N memory...not 4*N typical implementation...anyway..I guess you wanted to completely reorganise the interface..so I will make the commits and changes respectively..

File: pydatastructs/miscellaneous_data_structures/segment_tree.py (hunk 9 above)
Comment (Rajveer100) [reply pydatastructs/miscellaneous_data_structures/segment_tree.py:132]: Ok, will add it..

**Comment by Rajveer100** [comment]
Also, what about the previous implementation, which I initially made..so now that's not required anymore?

**Comment by czgdp1807** [comment]
> so now that's not required anymore (other than reference)

Yes, not required anymore.

**Comment by Rajveer100** [comment]
> Now, we need to add `RangeQueryDynamic` and `RangeQueryDynamicOneDimensionalArraySegmentTree` in `pydatastructs/miscellaneous_data_structures/algorithms.py`.
> 
> Lazy propagation can be added afterwards by adding a new class, something like, `OneDimensionalArraySegmentTreeWithLazyPropagation`. The reason is simple, we would need a new kind of node, i.e., `LazyTreeNode` which would have an attribute to store whether this node is required to be updated. I would recommend doing this in a follow up PR.

So now I have to do this right?

**Comment by czgdp1807** [comment]
You can if you want to.

**Comment by Rajveer100** [comment]
I mean, I was talking about the pr merge related way...because you have implemented the things in your way..would be hard to do it..so I have to switch to other issue..that's my doubt..as things have changed completely...

**Comment by Rajveer100** [comment]
Can always keep contributing in this anyway..

**Comment by czgdp1807** [comment]
In a follow up PR add, both the new APIs to docs.

**Comment by Rajveer100** [comment]
Hey Thanks...alot..meanwhile I am working on order statistic and interval trees also..Will commit them under their PRs..

**Comment by Rajveer100** [comment]
> In a follow up PR add, both the new APIs to docs.

Can you elaborate this?

**Comment by czgdp1807** [comment]
> order statistic

Could you please provide a reference for this?

**Comment by Rajveer100** [comment]
> > order statistic
> 
> Could you please provide a reference for this?

It's basically a self balancing tree like RB, AVL, B, van-embde-boas, etc...In the augmented form..where we stored two bits of extra information...for fetching osrank and osselect operations...please refer my C++ implementation for the same under "My Templates repo"...and here is a link for more info:

(CLRS pdf)
https://edutechlearners.com/download/Introduction_to_algorithms-3rd%20Edition.pdf

See the augmenting data structures part...before which red black trees is also given...

**Comment by czgdp1807** [comment]
Can you check if `is_order_statistic`, `select(i)` in https://pydatastructs.readthedocs.io/en/latest/pydatastructs/trees/binary_trees.html matches what you are trying to add?

**Comment by Rajveer100** [comment]
Yes it matched...but you have used AVL trees in your implementation?

**Comment by czgdp1807** [comment]
No. Any tree class can call `select`. Its generic. See the inheritance pattern. I think order statistic feature is already there. You can go ahead with interval trees though.

**Comment by Rajveer100** [comment]
So I think that I will start doing interval trees?...Since the repo already contains order statistic..And also this would be kinda nice...to have both kinds of trees...generally speaking RB Trees are most general and efficient also used in C++, Java..and their implementations of their DS..

**Comment by Rajveer100** [comment]
Cool...So I will get back to you directly under that issue..

**Comment by Rajveer100** [comment]
Thanks..

**Pull request merged**